numpy/__init__.pyi

```diff
@@ -1,8 +1,188 @@
 # very simple, just enough to start running tests
+#
+import builtins
+from typing import Any, Mapping, List, Optional, Tuple, Union
 
-class ndarray: pass
+from numpy.core._internal import _ctypes
+
+_Shape = Tuple[int, ...]
+
+class dtype:
+    names: Optional[Tuple[str, ...]]
+
+    @property
+    def alignment(self) -> int: ...
+
+    @property
+    def base(self) -> dtype: ...
+
+    @property
+    def byteorder(self) -> str: ...
+
+    @property
+    def char(self) -> str: ...
+
+    @property
+    def descr(self) -> List[Union[
+        Tuple[str, str],
+        Tuple[str, str, _Shape]]]: ...
+
+    @property
+    def fields(self) -> Optional[Mapping[
+        str,
+        Union[Tuple[dtype, int],
+              Tuple[dtype, int, Any]]]]: ...
+
+    @property
+    def flags(self) -> int: ...
+
+    @property
+    def hasobject(self) -> bool: ...
+
+    @property
+    def isbuiltin(self) -> int: ...
+
+    @property
+    def isnative(self) -> bool: ...
+
+    @property
+    def isalignedstruct(self) -> bool: ...
+
+    @property
+    def itemsize(self) -> int: ...
+
+    @property
+    def kind(self) -> str: ...
+
+    @property
+    def metadata(self) -> Optional[Mapping[str, Any]]: ...
+
+    @property
+    def name(self) -> str: ...
+
+    @property
+    def num(self) -> int: ...
+
+    @property
+    def shape(self) -> _Shape: ...
+
+    @property
+    def ndim(self) -> int: ...
+
+    @property
+    def subdtype(self) -> Optional[Tuple[dtype, _Shape]]: ...
+
+    def newbyteorder(self, new_order: str = ...) -> dtype: ...
+
+    # Leave str and type for end to avoid having to use `builtins.str`
+    # everywhere. See https://github.com/python/mypy/issues/3775
+    @property
+    def str(self) -> builtins.str: ...
+
+    @property
+    def type(self) -> builtins.type: ...
+
+
+_dtype_class = dtype      # for ndarray type
```

Comment:
Why does this have a duplicate name?

Reply:
Otherwise mypy gets confused when you to type the `dtype` ndarray attribute, e.g., `def dtype(self) -> dtype` refers to `ndarray.dtype` for the return value, not the `dtype` class.

Reply:
Can you fix it with an `import numpy as np`, and then using `np.dtype`?

Reply:
I don't know exactly how mypy handles circular imports, but I suspect that importing a module from itself would not work.